fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/limits/FDBRecordStoreByteLimitTest.java

```diff
@@ -0,0 +1,471 @@
+                    // Check that we stopped because of the byte scan limit.
+                    assertEquals(RecordCursor.NoNextReason.BYTE_LIMIT_REACHED, result.getNoNextReason());
+                }
+                // Assertion of claim (b)
+                assertThat(bytesScanned, lessThanOrEqualTo(2 * byteCountsByRecord.get(currentIndex)));
+                continuation = result.getContinuation().toBytes();
+            }
+        }
+
+        continuation = null;
+        i = 0;
+        while (i < byteCountsByRecord.size()) {
+            // If the limit is slightly too high we should scan the record, proceed to the next one (if it exists), and then stop.
+            final ExecuteProperties executeProperties = ExecuteProperties.newBuilder()
+                    .setScannedBytesLimit(byteCountsByRecord.get(i) + 1)
+                    .build();
+            try (RecordCursor<FDBQueriedRecord<Message>> cursor = recordStore.executeQuery(plan, continuation, executeProperties)) {
+                RecordCursorResult<FDBQueriedRecord<Message>> result = cursor.onNext().join();
+                assertTrue(result.hasNext());
+                continuation = result.getContinuation().toBytes();
+                result = cursor.onNext().join();
+
+                long bytesScanned = byteCounter.getBytesScanned(context);
+                // Assertion of claim (c)
+                assertThat(bytesScanned, greaterThanOrEqualTo(byteCountsByRecord.get(i)));
+                if (i < byteCountsByRecord.size() - 1) {
+                    // Assertion of claim (c)
+                    assertThat(bytesScanned, lessThanOrEqualTo(byteCountsByRecord.get(i) + byteCountsByRecord.get(i + 1)));
+                }
+
+                i++;
+                if (result.hasNext()) {
+                    result = cursor.onNext().join();
+                    assertFalse(result.hasNext());
+                }
+                context.getTimer().reset();
+            }
+        }
+    }
+
+    @ParameterizedTest(name = "plansByContinuation() [{index}] {0}")
+    @MethodSource("plans")
+    public void testPlansReturnSameRecordsRegardlessOfLimit(String description, boolean notUsed, RecordQueryPlan plan) throws Exception {
+        setupSimpleRecordStore();
+        final Function<FDBQueriedRecord<Message>, Long> getRecNo = r -> {
+            TestRecords1Proto.MySimpleRecord.Builder record = TestRecords1Proto.MySimpleRecord.newBuilder();
+            record.mergeFrom(r.getRecord());
+            return record.getRecNo();
+        };
+
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+            final List<Long> allAtOnce;
+            try (RecordCursor<FDBQueriedRecord<Message>> cursor = recordStore.executeQuery(plan)) {
+                allAtOnce = cursor.map(getRecNo).asList().get();
+            }
+
+            for (long byteLimit = 0; byteLimit < 1000; byteLimit += 10) {
```

Comment:
How long does this test take? I suspect there may be a way of testing this without running each query 100 times...

Reply:
Maybe every 100 bytes instead of every 10?